File: src/python/fru_cache.py

class FRUCache:
    def __init__(self, active_size=10, ghost_size=20, alpha=1.0, beta=1.0):
        self.active = {}  # key → (value, weight, last_tick)
        self.ghost = {}   # key → (weight, last_tick)
        self.tick = 0
        self.alpha = alpha
        self.beta = beta
        self.active_size = active_size
        self.ghost_size = ghost_size

    def _weight_increment(self, delta):
        return self.alpha * (delta ** self.beta)

    def get(self, key):
        self.tick += 1
        if key in self.active:
            val, weight, last_tick = self.active[key]
            delta = self.tick - last_tick
            new_weight = weight + self._weight_increment(delta)
            self.active[key] = (val, new_weight, self.tick)
            return val
        elif key in self.ghost:
            weight, last_tick = self.ghost[key]
            delta = self.tick - last_tick
            new_weight = weight + self._weight_increment(delta)
            self._promote(key, new_weight)
        return None

    def set(self, key, value):
        self.tick += 1
        if key in self.active:
            _, weight, last_tick = self.active[key]
            delta = self.tick - last_tick
            self.active[key] = (value, weight + self._weight_increment(delta), self.tick)
        else:
            if len(self.active) >= self.active_size:
                victim_key = min(self.active, key=lambda k: self.active[k][1])
                v_val, v_weight, v_tick = self.active[victim_key]
                self.ghost[victim_key] = (v_weight, v_tick)
                del self.active[victim_key]
            self.active[key] = (value, 1.0, self.tick)
            if len(self.ghost) > self.ghost_size:
                lru_key = min(self.ghost, key=lambda k: self.ghost[k][1])
                del self.ghost[lru_key]

    def _promote(self, key, new_weight):
        if len(self.active) >= self.active_size:
            victim_key = min(self.active, key=lambda k: self.active[k][1])
            v_val, v_weight, v_tick = self.active[victim_key]
            self.ghost[victim_key] = (v_weight, v_tick)
            del self.active[victim_key]
        self.active[key] = (None, new_weight, self.tick)
        if key in self.ghost:
            del self.ghost[key]

File: src/cpp/fru_cache.cpp

#include <unordered_map>
#include <map>
#include <limits>
#include <cmath>
#include <string>
#include <iostream>
#include <optional>

template<typename Key, typename Value>
class FRUCache {
public:
    FRUCache(size_t active_size, size_t ghost_size, double alpha = 1.0, double beta = 1.0)
        : active_size_(active_size), ghost_size_(ghost_size),
          alpha_(alpha), beta_(beta), tick_(0) {}

    std::optional<Value> get(const Key& key) {
        tick_++;
        auto it = active_.find(key);
        if (it != active_.end()) {
            double delta = tick_ - it->second.last_tick;
            it->second.weight += weight_increment(delta);
            it->second.last_tick = tick_;
            return it->second.value;
        }
        auto git = ghost_.find(key);
        if (git != ghost_.end()) {
            double delta = tick_ - git->second.last_tick;
            double new_weight = git->second.weight + weight_increment(delta);
            promote(key, new_weight);
        }
        return std::nullopt;
    }

    void set(const Key& key, const Value& value) {
        tick_++;
        auto it = active_.find(key);
        if (it != active_.end()) {
            double delta = tick_ - it->second.last_tick;
            it->second.weight += weight_increment(delta);
            it->second.last_tick = tick_;
            it->second.value = value;
            return;
        }

        if (active_.size() >= active_size_) {
            auto victim = std::min_element(active_.begin(), active_.end(),
                [](const auto& a, const auto& b) {
                    return a.second.weight < b.second.weight;
                });
            ghost_[victim->first] = {victim->second.weight, victim->second.last_tick};
            active_.erase(victim);
        }

        active_[key] = {value, 1.0, tick_};
        if (ghost_.size() > ghost_size_) {
            auto oldest = std::min_element(ghost_.begin(), ghost_.end(),
                [](const auto& a, const auto& b) {
                    return a.second.last_tick < b.second.last_tick;
                });
            ghost_.erase(oldest);
        }
    }

private:
    struct Entry {
        Value value;
        double weight;
        size_t last_tick;
    };

    struct GhostEntry {
        double weight;
        size_t last_tick;
    };

    double weight_increment(double delta) const {
        return alpha_ * std::pow(delta, beta_);
    }

    void promote(const Key& key, double new_weight) {
        if (active_.size() >= active_size_) {
            auto victim = std::min_element(active_.begin(), active_.end(),
                [](const auto& a, const auto& b) {
                    return a.second.weight < b.second.weight;
                });
            ghost_[victim->first] = {victim->second.weight, victim->second.last_tick};
            active_.erase(victim);
        }
        active_[key] = {Value(), new_weight, tick_};
        ghost_.erase(key);
    }

    std::unordered_map<Key, Entry> active_;
    std::unordered_map<Key, GhostEntry> ghost_;
    size_t active_size_;
    size_t ghost_size_;
    double alpha_;
    double beta_;
    size_t tick_;
};
